log4j-api/src/main/java/org/apache/logging/log4j/spi/LoggerRegistry.java

```diff
@@ -243,32 +232,30 @@ public boolean hasLogger(final String name, final Class<? extends MessageFactory
      * Registers the provided logger.
      * <b>Logger name and message factory parameters are ignored</b>, those will be obtained from the logger instead.
      *
-     * @param name ignored – kept for backward compatibility
-     * @param messageFactory ignored – kept for backward compatibility
+     * @param name a logger name
```

Comment:
There are two main reasons:

1. We want to make sure that `getLogger(name, mf)` returns the object that was inserted using `putIfAbsent(name, mf, logger)`. This way `getLogger` will not return `null` even if `(name, mf)` is different from `(logger.getName(), logger.getMessageFactory())`.
2. I want to use the `messageFactory` argument, so it does not become unreachable. This is not a very important argument, since `messageFactory` might become unreachable right after the `putIfAbsent` method call.

Reply:
> We want to make sure that `getLogger(name, mf)` returns the object that was inserted using `putIfAbsent(name, mf, logger)`. This way `getLogger` will not return `null` even if `(name, mf)` is different from `(logger.getName(), logger.getMessageFactory())`.

Can't we solve this problem by extracting `name` and `MF` from the `Logger` in `putIfAbsent()` too?

Reply:
What about registering a logger using both the coordinates provided by the caller of `putIfAbsent` and those inferred from the `Logger` instance? See https://github.com/apache/logging-log4j2/pull/3209/commits/2f1065d331af9d4df351d0fe41af80b49041e2b2

As far as I am concerned, the only valid reason for mismatching message factories might be provided by the proposed "metrics extension" (#2469)

Reply:
I am in favor of using the bare minimum input – which in this case is the passed `Logger` – as the single source of truth, but you can proceed as you wish.

Reply:
These 3 lines in `log4j-core` version `2.24.0` **must** work correctly and **never** return `null`:

https://github.com/apache/logging-log4j2/blob/c79ae325f6a21af45526c202f121bfced188613e/log4j-core/src/main/java/org/apache/logging/log4j/core/LoggerContext.java#L534-L536

Sure, if the `messageFactory` parameter is `null`, then the logger will be registered using `ParameterizedMessageFactory.INSTANCE` instead of the real one. However, at least, `getLogger(name, messageFactory)` will not return `null`, which is what would happen if I register the logger using `logger.getMessageFactory()`.